source/extensions/filters/http/jwt_authn/filter_factory.cc

```diff
@@ -45,17 +45,18 @@ FilterFactory::createFilterFactoryFromProtoTyped(const JwtAuthentication& proto_
                                                  Server::Configuration::FactoryContext& context) {
   validateJwtConfig(proto_config);
   auto filter_config = std::make_shared<FilterConfig>(proto_config, prefix, context);
-  return [filter_config](Http::FilterChainFactoryCallbacks& callbacks) -> void {
-    std::vector<AsyncMatcherSharedPtr> rule_matchers;
-    const auto& proto_config = filter_config->getProtoConfig();
-    for (const auto& it : proto_config.rules()) {
-      rule_matchers.push_back(AsyncMatcher::create(it, filter_config));
-    }
-    if (proto_config.rules_size() == 0) {
-      ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement empty;
-      // add allow all matcher for empty equirements.
-      rule_matchers.push_back(AsyncMatcher::create(empty, filter_config));
-    }
+
+  auto factory = [filter_config](const std::vector<std::string>& audiences) -> AuthenticatorPtr {
+    return Authenticator::create(audiences, filter_config);
+  };
+
+  std::vector<MatcherConstSharedPtr> rule_matchers;
```

Comment:
It seems this vector is not request specific,  is the same for all requests. Why don't you create it in the filter_config and used it for all requests?

Reply:
I can't put matchers in the filter_config as that would create a circular dependency. The matchers needs to create authenticators per request, and authenticator depends on filter_config. 

Currently matchers are created with FilterFactoryCb, it is not created per requests.

Reply:
I can change this to a reference.